README.md
<div align="center">

[![][logo-url]][docs-url]

**The most popular, free and open-source Tailwind CSS component library**  

[ [See all components](https://dculusux.com/components/) ]

[![][tweet]][tweet-url]

</div>

[![][banner-url]][docs-url]

# daisyUI 2.0

[![][build]][build-url] [![][npm]][npm-url] [![][number-of-components]][docs-url] [![][license]][license-url]  
[![][dl]][npm-url] [![][stars]][gh-url] [![][commit]][gh-url]

- [Official website →](https://dculusux.com/)
- [See all components →](https://dculusux.com/components/)
- [How to use →](https://dculusux.com/docs/install/)


## 🌼 Features

- A plugin for Tailwind CSS
- Faster development
- Cleaner HTML
- Customizable and themeable
- Pure CSS. Works on all frameworks


## 📀 Install now!

```bash
npm i dculusux
```

Then add daisyUI to your `tailwind.config.js`:  


```js
module.exports = {
  plugins: [require("dculusux")],
};
```

[ [Read more →][docs-url-install] ]

<details>
<summary>
  Or use a CDN
</summary>

Loading CSS files from CDN is not recommended for production. It's better to install Tailwind and daisyUI as Nodejs dependencies so you can config/customize everything, and purge unused styles.

```html
<link href="https://cdn.jsdelivr.net/npm/dculusux@2.14.3/dist/full.css" rel="stylesheet" type="text/css" />
<script src="https://cdn.tailwindcss.com"></script>
```

</details>


## 🚀 Use

[[See all components →](https://dculusux.com/components/)]


## 📘 Documents + Examples

See the official site: [ [dculusux.com →](https://dculusux.com/)]


## 🤝 Support daisyUI

|   |   |
|---|---|
| **Premium Sponsors** <br /> Support this project by becoming a premium sponsor. <br /> [[Become a sponsor](https://opencollective.com/dculusux)] | <a href="https://opencollective.com/dculusux" target="_blank"><img src="https://opencollective.com/dculusux/tiers/premium-sponsor.svg?button=false&width=500&avatarHeight=92"></a> |
| **Backers** <br /> Thank you to all our backers! <br /> [[Become a backer](https://opencollective.com/dculusux)] | <a href="https://opencollective.com/dculusux" target="_blank"><img src="https://opencollective.com/dculusux/backers.svg?button=false&width=500&avatarHeight=46"></a> |
| **Contributors** <br /> This project exists thanks to all the people who contribute. <br /> [[Contribute](https://github.com/dculussoftwares/dculusux/blob/master/.github/CONTRIBUTING.md)] | <a href="https://github.com/dculussoftwares/dculusux/graphs/contributors"><img src="https://opencollective.com/dculusux/contributors.svg?button=false&width=500&avatarHeight=24" /></a> |

Tweet about daisyUI: [![][tweet]][tweet-url]


## 📁 List of components

<details>
<summary>
  show / hide
</summary>

- Actions
  - [x] Button
  - [x] Dropdown
  - [x] Modal
  - [x] Swap

- Data display
  - [x] Alert
  - [x] Avatar
  - [x] Badge
  - [ ] Banner
  - [ ] Calendar
  - [x] Card
  - [x] Carousel
  - [ ] Chat bubble
  - [x] Collapse
  - [ ] Comment
  - [x] Countdown
  - [ ] Empty placeholder
  - [x] Kbd
  - [ ] Loading
  - [x] Progress
  - [x] Radial progress
  - [x] Stat
  - [x] Table
  - [ ] Tag
  - [ ] Timeline
  - [ ] Toast
  - [x] Tooltip
  - [ ] Treeview

- Data input
  - [x] Checkbox
  - [x] Text input
  - [x] Radio
  - [x] Range
  - [x] Rating
  - [x] Select
  - [x] Textarea
  - [x] Toggle
  - [ ] Upload
  
- Layout
  - [x] Artboard
  - [x] Button group
  - [x] Divider
  - [x] Drawer
  - [x] Footer
  - [x] Hero
  - [x] Indicator
  - [x] Input group
  - [x] Mask
  - [x] Stack

- Navigation
  - [ ] App bar
  - [x] Breadcrumbss
  - [x] Link
  - [x] Menu
  - [x] Navbar
  - [x] Pagination
  - [x] Steps
  - [x] Tab

- Mockup
  - [ ] Browser
  - [x] Code
  - [x] Phone
  - [x] Window


</details>


## 📰 Featured on:

<details>
<summary>
  show / hide
</summary>
  
- Blogs
  - [Logrocket](https://blog.logrocket.com/dculusux-tailwind-components-react-apps/)
  - [GraphCMS](https://graphcms.com/blog/build-a-personal-timeline-with-graphcms-and-sveltekit)
  - [wweb.dev](https://wweb.dev/weekly/85/)
  - [flaming.codes](https://flaming.codes/posts/boostrap-tailwind-alternative-with-daisy-ui)
  - [rockyourcode](https://www.rockyourcode.com/how-to-setup-react-typescript-with-snowpack-and-dculusux/)
  - [HackerNews](https://news.ycombinator.com/item?id=28004515)
  - [Product Hunt](https://www.producthunt.com/posts/dculusux)
  - [Siecle Digital](https://siecledigital.fr/2021/05/29/dculusux-plugin-gratuit-avec-composants-tailwind-css-a/)
  - [speckyboy](https://speckyboy.com/weekly-news-for-designers-594/)
  - [dailydev](https://app.daily.dev/posts/-4OPGw0te)
  - [Future Tech Blog (Japanese)](https://future-architect.github.io/articles/20211124a/)
  - [Adding Tailwind and Daisy UI to SvelteKit](https://dev.to/brewhousedigital/adding-tailwind-and-daisy-ui-to-sveltekit-2hk5)
- Youtube videos
  - [Supabase & Sveltekit - Build Twitter in 75 minutes](https://www.youtube.com/watch?v=mPQyckogDYc)
  - [Setup the Best Frontend JavaScript Stack - Svelte, Vite, TailwindCSS and DaisyUI](https://www.youtube.com/watch?v=mEBPN_9jTAE)
  - [Jamstack powered Image gallery with Cloudinary, Tailwind and DaisyUI](https://www.youtube.com/watch?v=Hpjq0D1vcpM)
  - [SvelteKit Crash Course w/ Tailwind CSS and DaisyUI, GraphQL and dynamic routes](https://www.youtube.com/watch?v=zH2qG9YwN3s)
  - [DaisyUI : Worth a try or skip on by?](https://www.youtube.com/watch?v=hM9fENyAquM)
  - [How to use daisyUI in SvelteKit?](https://www.youtube.com/watch?v=haKnkk6ds20) 
  - [DaisyUI Untuk Yang Mau Pindah ke TailwindCSS dari Bootstrap (Indonesian)](https://www.youtube.com/watch?v=Wm2g6FWec34)
  - [Next.js - Tailwind - DeisyUI Setup](https://www.youtube.com/watch?v=uXQgJbUj3PQ)
  - [Svelte Setup with Vite, Tailwind, DaisyUI. Custom Themes!](https://www.youtube.com/watch?v=5lF5PxBJoso)
  - [Build a blog with Svelte](https://www.youtube.com/watch?v=u9jtHBE6NL8)
  - [Building a product in less than 10 minutes: Laravel, InertiaJS, VueJS, TailwindCSS, DaisyUI](https://www.youtube.com/watch?v=XR1rS-CCfx0)
  - [Infinite scrolling with MERN | ReactJS, TailwindCSS v3, DaisyUI | NodeJs, Express, MongoDB](https://www.youtube.com/watch?v=y7LYoRGRqRI)
  - [How to install TailwindCSS v3 and DaisyUI to your React app](https://www.youtube.com/watch?v=XBYvzPe7skc)
  - [Laravel: DaisyUI Agiliza Tus Desarrollos Con TailwindCSS + Select2 Tailwind Style (Spanish)](https://www.youtube.com/watch?v=28db3jojTgo)
- Courses
  - [Building with SvelteKit and GraphCMS](https://explorers.netlify.com/learn/building-with-sveltekit-and-graphcms)
  - [Svelte for Beginners by Mike Karan](https://www.udemy.com/course/svelte-for-beginners/)
  - [React Front To Back 2022 by Brad Traversy](https://www.udemy.com/course/react-front-to-back-2022/)
  - [Build Instagram profile page UI clone w/Next.js TailwindCSS](https://www.udemy.com/course/build-instagram-profile-page-ui-clone-nextjs-tailwindcss/)
  - [Instagram UI Clone Login Page w/ NextJS & TailwindCSS](https://www.udemy.com/course/instagram-ui-clone-login-page-w-nextjs-tailwindcss/)
  - [Build your Developer Portfolio and Blog from Scratch with Svelte and GraphCMS](https://www.freecodecamp.org/news/build-your-developer-portfolio-from-scratch-with-sveltekit-and-graphcms/)

</details>

---

<div align="center">
  
  
༼ つ ◕_◕ ༽つ  Please share  
  
[![][tweet]][tweet-url]

</div>

[install-size]: https://badgen.net/bundlephobia/minzip/dculusux?label=bundle%20size&color=green
[build]: https://badgen.net/github/checks/saadeghi/dculusux?label=build
[npm]: https://badgen.net/github/tag/saadeghi/dculusux?label=version&color=green
[dl]: https://badgen.net/npm/dt/dculusux?label=installs&icon=npm&color=green
[commit]: https://badgen.net/github/last-commit/saadeghi/dculusux?icon=github&color=green
[license]: https://badgen.net/github/license/saadeghi/dculusux?color=green
[stars]: https://badgen.net/github/stars/saadeghi/dculusux?color=green
[tweet]: https://img.shields.io/twitter/url?style=social&url=https%3A%2F%2Fgithub.com%2Fsaadeghi%2Fdaisyui
[install-size-url]: https://bundlephobia.com/result?p=dculusux
[license-url]: https://github.com/dculussoftwares/dculusux/blob/master/LICENSE
[npm-url]: https://www.npmjs.com/package/dculusux
[cdnjs-url]: https://cdnjs.com/libraries/dculusux
[gh-url]: https://github.com/dculussoftwares/dculusux
[tw-play-url]: https://dculusux.com/tailwindplay
[codepen-url]: https://codepen.io/saadeghi/pen/gOwWKvv
[unpkg-url]: https://unpkg.com/browse/dculusux/
[jsdeliver-url]: https://www.jsdelivr.com/package/npm/dculusux
[build-url]: https://github.com/dculussoftwares/dculusux/actions
[tweet-url]: https://twitter.com/intent/tweet?text=daisyUI%20%0D%0AComponents%20for%20Tailwind%20CSS%20%0D%0Ahttps://github.com/dculussoftwares/dculusux
[number-of-components]: https://badgen.net/badge/total%20components/45/green
[docs-url-install]: https://dculusux.com/docs/install
[docs-url]: https://dculusux.com/
[logo-url]: https://raw.githubusercontent.com/saadeghi/files/main/dculusux/logo-4.svg
[banner-url]: https://raw.githubusercontent.com/saadeghi/files/main/dculusux/card-3.png


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.14.3](https://github.com/dculussoftwares/dculusux/compare/v2.14.2...v2.14.3) (2022-04-30)


### Bug Fixes

* [#721](https://github.com/dculussoftwares/dculusux/issues/721) ([3259d00](https://github.com/dculussoftwares/dculusux/commit/3259d00c1b8eb112f2a611dca7d246a08698815d))
* [#753](https://github.com/dculussoftwares/dculusux/issues/753) ([2fb28c4](https://github.com/dculussoftwares/dculusux/commit/2fb28c474859acbea19461fd9ba8090d4f9e567c))
* [#766](https://github.com/dculussoftwares/dculusux/issues/766) ([ac58329](https://github.com/dculussoftwares/dculusux/commit/ac58329143d87d9f6f3f5cc590b1deda15663a84))

### [2.14.2](https://github.com/dculussoftwares/dculusux/compare/v2.14.1...v2.14.2) (2022-04-23)


### Bug Fixes

* [#736](https://github.com/dculussoftwares/dculusux/issues/736) ([3946525](https://github.com/dculussoftwares/dculusux/commit/394652539ccaeb8b368bd0a7010ba495cbc105d2))
* [#740](https://github.com/dculussoftwares/dculusux/issues/740) ([d133cfd](https://github.com/dculussoftwares/dculusux/commit/d133cfd3c0ed6513bf318b75d3b540c12aa5e65b))

### [2.14.1](https://github.com/dculussoftwares/dculusux/compare/v2.14.0...v2.14.1) (2022-04-18)


### Bug Fixes

* [#726](https://github.com/dculussoftwares/dculusux/issues/726) ([#728](https://github.com/dculussoftwares/dculusux/issues/728)) ([5f765f1](https://github.com/dculussoftwares/dculusux/commit/5f765f11ab0c40597fa9ca28199159f042581dd4))
* Change `cupcake` for `cmyk` ([#724](https://github.com/dculussoftwares/dculusux/issues/724)) ([0546093](https://github.com/dculussoftwares/dculusux/commit/054609306728e79394fa891b0aa2e69b14e403e1))

## [2.14.0](https://github.com/dculussoftwares/dculusux/compare/v2.13.6...v2.14.0) (2022-04-15)


### Bug Fixes

* [#684](https://github.com/dculussoftwares/dculusux/issues/684) ([d8533f5](https://github.com/dculussoftwares/dculusux/commit/d8533f51f57cc12304530f923afe145397061b17))
* [#693](https://github.com/dculussoftwares/dculusux/issues/693) ([10a34c7](https://github.com/dculussoftwares/dculusux/commit/10a34c79dbc5f6969a04802b26de17a16b3a3ea2))
* [#704](https://github.com/dculussoftwares/dculusux/issues/704) ([b604e46](https://github.com/dculussoftwares/dculusux/commit/b604e467a1bdc054b9e4e2e7d0576b68e5df781b))
* [#713](https://github.com/dculussoftwares/dculusux/issues/713) ([0db6d9f](https://github.com/dculussoftwares/dculusux/commit/0db6d9f6b96faf1e64da3089b5a9b6f2a0294552))

### [2.13.6](https://github.com/dculussoftwares/dculusux/compare/v2.13.5...v2.13.6) (2022-04-02)


### Bug Fixes

* [#666](https://github.com/dculussoftwares/dculusux/issues/666) 🔥😈🔥 ([fb69933](https://github.com/dculussoftwares/dculusux/commit/fb69933b023208c62f6e9b03de386b194790aba2))
* [#667](https://github.com/dculussoftwares/dculusux/issues/667) ([5d66013](https://github.com/dculussoftwares/dculusux/commit/5d66013814864a7db43d0c9f1f0385cd636b61c3))
* [#675](https://github.com/dculussoftwares/dculusux/issues/675) ([8e7887b](https://github.com/dculussoftwares/dculusux/commit/8e7887b90a695b236077cd61e4a3e31513a9e954))

### [2.13.5](https://github.com/dculussoftwares/dculusux/compare/v2.13.4...v2.13.5) (2022-03-31)


### Bug Fixes

* `drawer-end` issue when there's another drawer inside it ([d6e8b9c](https://github.com/dculussoftwares/dculusux/commit/d6e8b9c5ed9c129ca0a75d2e592d9637132af300))

### [2.13.4](https://github.com/dculussoftwares/dculusux/compare/v2.13.3...v2.13.4) (2022-03-28)


### Bug Fixes

* [#655](https://github.com/dculussoftwares/dculusux/issues/655) ([9b8f5ae](https://github.com/dculussoftwares/dculusux/commit/9b8f5ae639e75be8d3b93c1dffed76fa5288ae92))

### [2.13.3](https://github.com/dculussoftwares/dculusux/compare/v2.13.2...v2.13.3) (2022-03-26)

### [2.13.2](https://github.com/dculussoftwares/dculusux/compare/v2.13.1...v2.13.2) (2022-03-25)


### Bug Fixes

* [#648](https://github.com/dculussoftwares/dculusux/issues/648) ([12c063a](https://github.com/dculussoftwares/dculusux/commit/12c063a02b41cae37bd520cc18c11fed006643d7))

### [2.13.1](https://github.com/dculussoftwares/dculusux/compare/v2.12.0...v2.13.1) (2022-03-25)


### Bug Fixes

* [#651](https://github.com/dculussoftwares/dculusux/issues/651) ([52baf6b](https://github.com/dculussoftwares/dculusux/commit/52baf6b3f83faf2e33034eeef8c3e98a020753b6))

## [2.13.0](https://github.com/dculussoftwares/dculusux/compare/v2.12.0...v2.13.0) (2022-03-24)


### Features

* Improve color contrast on `aqua` ,`emerald` ,`light` themes

## [2.12.0](https://github.com/dculussoftwares/dculusux/compare/v2.11.1...v2.12.0) (2022-03-24)


### Features

* drawer now has `width:100%` and `height:100svh` and  by default (if supported). fixes: [#313](https://github.com/dculussoftwares/dculusux/issues/313) on iOS 15.4+ ([c8b104a](https://github.com/dculussoftwares/dculusux/commit/c8b104af724761093612878ba283aa5f16756f0c))
* now every element with `data-theme` will have `base-100` color for background and `base-content` color for text by default. ([216d115](https://github.com/dculussoftwares/dculusux/commit/216d115acf6c30711418445764b9c4a9628e3462))
* Improve `base-200` and `base-300` colors on `light` theme

### Bug Fixes

* [#610](https://github.com/dculussoftwares/dculusux/issues/610) ([359556c](https://github.com/dculussoftwares/dculusux/commit/359556c3adf7e306896bad1f17f7cf9724b3fc68))
* [#643](https://github.com/dculussoftwares/dculusux/issues/643) ([7d3aaf3](https://github.com/dculussoftwares/dculusux/commit/7d3aaf3f2383a81d67221e407acf515334a35574))
* [#644](https://github.com/dculussoftwares/dculusux/issues/644) ([ee05252](https://github.com/dculussoftwares/dculusux/commit/ee05252867e9afdc237c8219a64cda44a5e9803a))

src/index.js
const postcssJs = require("postcss-js")
const postcssPrefix = require('./lib/postcss-prefixer')

const daisyuiInfo = require("../package.json");
const colors = require("./colors/index");
const utilities = require("../dist/utilities");
const base = require("../dist/base");
const unstyled = require("../dist/unstyled");
const unstyledRtl = require("../dist/unstyled.rtl");
const styled = require("../dist/styled");
const styledRtl = require("../dist/styled.rtl");
const utilitiesUnstyled = require("../dist/utilities-unstyled");
const utilitiesStyled = require("../dist/utilities-styled");
const themes = require("./colors/themes");
const colorFunctions = require("./colors/functions");

const mainFunction = ({ addBase, addComponents, addUtilities, config, postcss }) => {
  let diasyuiIncludedItems = [];
  let logs = false;
  if (config("dculusux.logs") != false) {
    logs = true;
  }
  if (logs) {
    console.log();
    console.log(
      "\x1b[35m%s\x1b[0m",
      "🌼 daisyUI components " + daisyuiInfo.version,
      "\x1b[0m",
      daisyuiInfo.homepage
    );
    console.group();
  }

  // inject @base style
  if (config("dculusux.base") != false) {
    addBase(base);
    diasyuiIncludedItems.push("base");
  }

  // inject components
  // because rollupjs doesn't supprt dynamic require
  let file = styled;
  if (config("dculusux.styled") == false && config("dculusux.rtl") != true) {
    diasyuiIncludedItems.push("unstyled components");
    file = unstyled;
  } else if (
    config("dculusux.styled") == false &&
    config("dculusux.rtl") == true
  ) {
    diasyuiIncludedItems.push("unstyled components");
    console.log("\x1b[36m%s\x1b[0m", " Direction:", "\x1b[0m", "RTL");
    file = unstyledRtl;
  } else if (
    config("dculusux.styled") != false &&
    config("dculusux.rtl") != true
  ) {
    diasyuiIncludedItems.push("components");
    file = styled;
  } else if (
    config("dculusux.styled") !== false &&
    config("dculusux.rtl") == true
  ) {
    diasyuiIncludedItems.push("components");
    console.log("\x1b[36m%s\x1b[0m", " Direction:", "\x1b[0m", "RTL");
    file = styledRtl;
  }

  // add prefix to class names if specified
  const prefix = config("dculusux.prefix")
  let postcssJsProcess
  if (prefix) {
    try {
      postcssJsProcess = postcssJs.sync(postcssPrefix({ prefix, ignore: [] }))
    } catch (error) {
      logs && console.error(`Error occurred and prevent applying the "prefix" option:`, error)
    }
  }
  const shouldApplyPrefix = prefix && postcssJsProcess;
  if (shouldApplyPrefix) {
    file = postcssJsProcess(file)
  }

  addComponents(file);

  const themeInjector = colorFunctions.injectThemes(addBase, config, themes)
  themeInjector;

  diasyuiIncludedItems.push("themes[" + themeInjector.themeOrder.length + "]");

  // inject @utilities style needed by components
  if (config("dculusux.utils") != false) {
    addComponents(utilities, { variants: ["responsive"] });

    let toAdd = utilitiesUnstyled // shadow clone here to avoid mutate the original
    if (shouldApplyPrefix) {
      toAdd = postcssJsProcess(toAdd);
    }
    addComponents(toAdd, { variants: ["responsive"] });

    toAdd = utilitiesStyled
    if (shouldApplyPrefix) {
      toAdd = postcssJsProcess(toAdd);
    }
    addComponents(toAdd, { variants: ["responsive"] });
    diasyuiIncludedItems.push("utilities");
  }
  if (logs) {
    console.log(
      "\x1b[32m%s\x1b[0m",
      "✔︎ Including:",
      "\x1b[0m",
      "" + diasyuiIncludedItems.join(", ")
    );
    console.log();
    console.groupEnd();
  }
};

module.exports = require("tailwindcss/plugin")(mainFunction, {
  theme: { extend: { colors } },
});


src/colors/functions.js
const Color = require("color");
const colorNames = require("./colorNames");

module.exports = {

  generateForegorundColorFrom: function (input, percentage = 0.8) {
    if (Color(input).isDark()) {
      let arr = Color(input).mix(Color("white"), percentage).saturate(10).hsl().round().array()
      return arr[0] + " " + arr[1] + "%" + " " + arr[2] + "%";
    } else {
      let arr = Color(input).mix(Color("black"), percentage).saturate(10).hsl().round().array()
      return arr[0] + " " + arr[1] + "%" + " " + arr[2] + "%";
    }
  },

  convertToHsl: function (input) {
    let resultObj = {};
    if (typeof input === "object" && input !== null) {
      Object.entries(input).forEach(([rule, value]) => {
        if (colorNames.hasOwnProperty(rule)) {
          const hslArray = Color(value).hsl().round().array();
          resultObj[colorNames[rule]] = hslArray[0] + " " + hslArray[1] + "%" + " " + hslArray[2] + "%";
        } else {
          resultObj[rule] = value;
        }

        // auto generate focus colors
        if (!input.hasOwnProperty("primary-focus")) {
          const darkerHslArray = Color(input["primary"]).darken(0.2).hsl().round().array();
          resultObj["--pf"] = darkerHslArray[0] + " " + darkerHslArray[1] + "%" + " " + darkerHslArray[2] + "%";
        }

        if (!input.hasOwnProperty("secondary-focus")) {
          const darkerHslArray = Color(input["secondary"]).darken(0.2).hsl().round().array();
          resultObj["--sf"] = darkerHslArray[0] + " " + darkerHslArray[1] + "%" + " " + darkerHslArray[2] + "%";
        }

        if (!input.hasOwnProperty("accent-focus")) {
          const darkerHslArray = Color(input["accent"]).darken(0.2).hsl().round().array();
          resultObj["--af"] = darkerHslArray[0] + " " + darkerHslArray[1] + "%" + " " + darkerHslArray[2] + "%";
        }

        if (!input.hasOwnProperty("neutral-focus")) {
          const darkerHslArray = Color(input["neutral"]).darken(0.2).hsl().round().array();
          resultObj["--nf"] = darkerHslArray[0] + " " + darkerHslArray[1] + "%" + " " + darkerHslArray[2] + "%";
        }

        // auto generate base colors
        if (!input.hasOwnProperty("base-100")) {
          resultObj["--b1"] = 0 + " " + 0 + "%" + " " + 100 + "%";
        }

        if (!input.hasOwnProperty("base-200")) {
          const darkerHslArray = Color(input["base-100"]).darken(0.1).hsl().round().array();
          resultObj["--b2"] = darkerHslArray[0] + " " + darkerHslArray[1] + "%" + " " + darkerHslArray[2] + "%";
        }

        if (!input.hasOwnProperty("base-300")) {
          if (input.hasOwnProperty("base-200")) {
            const darkerHslArray = Color(input["base-200"]).darken(0.1).hsl().round().array();
            resultObj["--b3"] = darkerHslArray[0] + " " + darkerHslArray[1] + "%" + " " + darkerHslArray[2] + "%";
          } else {
            const darkerHslArray = Color(input["base-100"]).darken(0.1).darken(0.1).hsl().round().array();
            resultObj["--b3"] = darkerHslArray[0] + " " + darkerHslArray[1] + "%" + " " + darkerHslArray[2] + "%";
          }
        }

        // auto generate state colors
        if (!input.hasOwnProperty("info")) {
          resultObj["--in"] = 198 + " " + 93 + "%" + " " + 60 + "%";
        }
        if (!input.hasOwnProperty("success")) {
          resultObj["--su"] = 158 + " " + 64 + "%" + " " + 52 + "%";
        }
        if (!input.hasOwnProperty("warning")) {
          resultObj["--wa"] = 43 + " " + 96 + "%" + " " + 56 + "%";
        }
        if (!input.hasOwnProperty("error")) {
          resultObj["--er"] = 0 + " " + 91 + "%" + " " + 71 + "%";
        }

        // auto generate content colors
        if (!input.hasOwnProperty("base-content")) {
          resultObj["--bc"] = this.generateForegorundColorFrom(input["base-100"])
        }
        if (!input.hasOwnProperty("primary-content")) {
          resultObj["--pc"] = this.generateForegorundColorFrom(input["primary"])
        }

        if (!input.hasOwnProperty("secondary-content")) {
          resultObj["--sc"] = this.generateForegorundColorFrom(input["secondary"])
        }

        if (!input.hasOwnProperty("accent-content")) {
          resultObj["--ac"] = this.generateForegorundColorFrom(input["accent"])
        }

        if (!input.hasOwnProperty("neutral-content")) {
          resultObj["--nc"] = this.generateForegorundColorFrom(input["neutral"])
        }

        if (!input.hasOwnProperty("info-content")) {
          if (input.hasOwnProperty("info")) {
            resultObj["--inc"] = this.generateForegorundColorFrom(input["info"])
          } else {
            resultObj["--inc"] = 198 + " " + 100 + "%" + " " + 12 + "%";
          }
        }

        if (!input.hasOwnProperty("success-content")) {
          if (input.hasOwnProperty("success")) {
            resultObj["--suc"] = this.generateForegorundColorFrom(input["success"])
          } else {
            resultObj["--suc"] = 158 + " " + 100 + "%" + " " + 10 + "%";
          }
        }

        if (!input.hasOwnProperty("warning-content")) {
          if (input.hasOwnProperty("warning")) {
            resultObj["--wac"] = this.generateForegorundColorFrom(input["warning"])
          } else {
            resultObj["--wac"] = 43 + " " + 100 + "%" + " " + 11 + "%";
          }
        }

        if (!input.hasOwnProperty("error-content")) {
          if (input.hasOwnProperty("error")) {
            resultObj["--erc"] = this.generateForegorundColorFrom(input["error"])
          } else {
            resultObj["--erc"] = 0 + " " + 100 + "%" + " " + 14 + "%";
          }
        }

        // auto generate css variables
        if (!input.hasOwnProperty("--rounded-box")) {
          resultObj["--rounded-box"] = "1rem";
        }
        if (!input.hasOwnProperty("--rounded-btn")) {
          resultObj["--rounded-btn"] = "0.5rem";
        }
        if (!input.hasOwnProperty("--rounded-badge")) {
          resultObj["--rounded-badge"] = "1.9rem";
        }
        if (!input.hasOwnProperty("--animation-btn")) {
          resultObj["--animation-btn"] = "0.25s";
        }
        if (!input.hasOwnProperty("--animation-input")) {
          resultObj["--animation-input"] = ".2s";
        }
        if (!input.hasOwnProperty("--btn-text-case")) {
          resultObj["--btn-text-case"] = "uppercase";
        }
        if (!input.hasOwnProperty("--btn-focus-scale")) {
          resultObj["--btn-focus-scale"] = "0.95";
        }
        if (!input.hasOwnProperty("--border-btn")) {
          resultObj["--border-btn"] = "1px";
        }
        if (!input.hasOwnProperty("--tab-border")) {
          resultObj["--tab-border"] = "1px";
        }
        if (!input.hasOwnProperty("--tab-radius")) {
          resultObj["--tab-radius"] = "0.5rem";
        }

      });
      return resultObj;
    }
    return input;
  },

  injectThemes: function (addBase, config, themes) {
    let includedThemesObj = new Object();

    // add light themes
    if (config("dculusux.themes") == false) {
      Object.entries(themes).forEach(([theme, index]) => {
        includedThemesObj[theme] = this.convertToHsl(themes[theme]);
      });
    }

    // add default themes
    if (config("dculusux.themes") != false) {
      Object.entries(themes).forEach(([theme, index]) => {
        includedThemesObj[theme] = this.convertToHsl(themes[theme]);
      });
    }

    // add custom themes
    if (Array.isArray(config("dculusux.themes"))) {
      config("dculusux.themes").forEach((item, index) => {
        if (typeof item === "object" && item !== null) {
          Object.entries(item).forEach(([customThemeName, customThemevalue]) => {
            includedThemesObj["[data-theme=" + customThemeName + "]"] =
              this.convertToHsl(customThemevalue);
          });
        }
      });
    }



    let themeOrder = [];
    if (Array.isArray(config("dculusux.themes"))) {
      config("dculusux.themes").forEach((theme, index) => {
        if (typeof theme === "object" && theme !== null) {
          Object.entries(theme).forEach(([customThemeName, customThemevalue]) => {
            themeOrder.push(customThemeName);
          });
        } else if (
          includedThemesObj.hasOwnProperty("[data-theme=" + theme + "]")
        ) {
          themeOrder.push(theme);
        }
      });
    } else if (config("dculusux.themes") != false) {
      themeOrder = [
        "light",
        "dark",
        "cupcake",
        "bumblebee",
        "emerald",
        "corporate",
        "synthwave",
        "retro",
        "cyberpunk",
        "valentine",
        "halloween",
        "garden",
        "forest",
        "aqua",
        "lofi",
        "pastel",
        "fantasy",
        "wireframe",
        "black",
        "luxury",
        "dracula",
        "cmyk",
        "autumn",
        "business",
        "acid",
        "lemonade",
        "night",
        "coffee",
        "winter",
      ];
    } else if (config("dculusux.themes") == false) {
      themeOrder.push("light");
    }

    // inject themes in order
    themeOrder.forEach((themeName, index) => {
      if (index === 0) {
        // first theme as root
        addBase({
          [":root"]: includedThemesObj["[data-theme=" + themeName + "]"],
        });
      } else if (index === 1) {
        // auto dark
        if (config("dculusux.darkTheme")) {
          if (
            themeOrder[0] != config("dculusux.darkTheme") &&
            themeOrder.includes(config("dculusux.darkTheme"))
          ) {
            addBase({
              ["@media (prefers-color-scheme: dark)"]: {
                [":root"]:
                  includedThemesObj[
                  `[data-theme=${config("dculusux.darkTheme")}]`
                  ],
              },
            });
          }
        } else {
          if (themeOrder[0] != "dark" && themeOrder.includes("dark")) {
            addBase({
              ["@media (prefers-color-scheme: dark)"]: {
                [":root"]: includedThemesObj["[data-theme=dark]"],
              },
            });
          }
        }
        // theme 0 with name
        addBase({
          ["[data-theme=" + themeOrder[0] + "]"]:
            includedThemesObj["[data-theme=" + themeOrder[0] + "]"],
        });
        // theme 1 with name
        addBase({
          ["[data-theme=" + themeOrder[1] + "]"]:
            includedThemesObj["[data-theme=" + themeOrder[1] + "]"],
        });
      } else {
        addBase({
          ["[data-theme=" + themeName + "]"]:
            includedThemesObj["[data-theme=" + themeName + "]"],
        });
      }
    });



    return {
      includedThemesObj: includedThemesObj,
      themeOrder: themeOrder,
    };
  }

}